internal/handler/msg_find.go

```diff
@@ -128,41 +126,57 @@ func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, er
 		return nil, lazyerrors.Error(err)
 	}
 
-	// Combine iterators chain and closer into a cursor to pass around.
-	// The context will be canceled when client disconnects or after maxTimeMS.
-	cursor := h.cursors.NewCursor(ctx, iterator.WithClose(iter, closer.Close), &cursor.NewParams{
+	t := cursor.Normal
+
+	if params.Tailable {
+		t = cursor.Tailable
+	}
+
+	if params.AwaitData {
+		t = cursor.TailableAwait
+	}
+
+	c := h.cursors.NewCursor(ctx, iter, &cursor.NewParams{
+		Data: &findCursorData{
+			coll:       coll,
+			qp:         qp,
+			findParams: params,
+		},
 		DB:           params.DB,
 		Collection:   params.Collection,
 		Username:     username,
+		Type:         t,
 		ShowRecordID: params.ShowRecordId,
 	})
 
-	cursorID := cursor.ID
+	cursorID := c.ID
 
-	firstBatchDocs, err := iterator.ConsumeValuesN(cursor, int(params.BatchSize))
+	docs, err := iterator.ConsumeValuesN(c, int(params.BatchSize))
 	if err != nil {
 		return nil, lazyerrors.Error(err)
 	}
 
 	h.L.Debug(
-		"Got first batch", zap.Int64("cursor_id", cursorID),
-		zap.Int("count", len(firstBatchDocs)), zap.Int64("batch_size", params.BatchSize),
+		"Got first batch", zap.Int64("cursor_id", cursorID), zap.Stringer("type", c.Type),
+		zap.Int("count", len(docs)), zap.Int64("batch_size", params.BatchSize),
 		zap.Bool("single_batch", params.SingleBatch),
 	)
 
-	firstBatch := types.MakeArray(len(firstBatchDocs))
-	for _, doc := range firstBatchDocs {
-		firstBatch.Append(doc)
-	}
+	if params.SingleBatch || len(docs) < int(params.BatchSize) {
+		c.Close()
 
-	if params.SingleBatch || firstBatch.Len() < int(params.BatchSize) {
-		// support tailable cursors
-		// TODO https://github.com/FerretDB/FerretDB/issues/2283
+		// It is not entirely clear if we should do that; more tests are needed.
+		if c.Type != cursor.Normal {
+			h.cursors.CloseAndRemove(c)
+		}
```

Comment:
Hm…

Reply:
All tests also pass if I change `!=` to `==`, so yeah, we need to add a test for that

Reply:
I've added this one: https://github.com/FerretDB/FerretDB/pull/3859/files#diff-ac456a92d3f0b992305fd1927bcc41d7c401a1d4622bcaddbee6e597605761e9R248 

It's still not dependent on this line, although I couldn't find any way to test it. Find won't return you any cursorID so you're not able to track it